From 340f56a110cbe5c4717e71edb6f48b44b828eb79 Mon Sep 17 00:00:00 2001
From: Antonio Gomes <tonikitoo@igalia.com>
Date: Wed, 31 Aug 2016 09:51:16 +0200
Subject: [PATCH 1/5] Make DesktopFactoryOzone instances to respect
 --ozone-platform

In order to support "use_ozone=1 chromeos=0" builds,
DesktopFactoryOzone class needs to respect the Ozone platform
backend that is going to run.

For instance, ui::OzonePlatform class (see ui/ozone/) accomplishes
this using the logic provided by ui::OzoneObject.
In short, OzoneObject allows OzonePlatform instances (e.g. wayland,
GDM, x11) to get selected at runtime, obeying either --ozone-platform
command line or using a default value to select the platform backend.

This patch extends this logic to DesktopFactoryOzone class, as a first
step towards supporting "use_ozone=1 chromeos=0" builds.
DesktopFactoryOzoneWayland is added (with dummy implementations) as
a sample implementation and starting point.

BUG=295089

Amend from Fred Wang <fwang@igalia.com>:

ui/view/BUILD.gn: Protect Ozone-specific part with use_ozone conditional.
---
 ui/views/BUILD.gn                                  | 33 +++++++++++++++++
 ui/views/DEPS                                      |  3 ++
 .../widget/desktop_aura/desktop_factory_ozone.cc   | 22 ++++++++----
 .../widget/desktop_aura/desktop_factory_ozone.h    |  3 --
 .../desktop_aura/desktop_factory_ozone_wayland.cc  | 23 ++++++++++++
 .../desktop_aura/desktop_factory_ozone_wayland.h   | 41 ++++++++++++++++++++++
 6 files changed, 116 insertions(+), 9 deletions(-)
 create mode 100644 ui/views/widget/desktop_aura/desktop_factory_ozone_wayland.cc
 create mode 100644 ui/views/widget/desktop_aura/desktop_factory_ozone_wayland.h

diff --git a/ui/views/BUILD.gn b/ui/views/BUILD.gn
index c78b7aa..0f35ca3 100644
--- a/ui/views/BUILD.gn
+++ b/ui/views/BUILD.gn
@@ -12,6 +12,11 @@ config("flags") {
   defines = [ "TOOLKIT_VIEWS=1" ]
 }
 
+if (use_ozone) {
+  import("//ui/ozone/ozone.gni")
+  constructor_list_cc_file = "$target_gen_dir/desktop_factory_ozone_list.cc"
+}
+
 component("views") {
   all_dependent_configs = [ ":flags" ]
   sources = [
@@ -615,6 +620,16 @@ component("views") {
           "widget/desktop_aura/desktop_screen_ozone.cc",
           "widget/desktop_aura/desktop_window_tree_host_ozone.cc",
         ]
+        deps += [
+          ":generate_constructor_list",
+        ]
+        sources += [ constructor_list_cc_file ]
+        if (ozone_platform_wayland) {
+          sources += [
+            "widget/desktop_aura/desktop_factory_ozone_wayland.cc",
+            "widget/desktop_aura/desktop_factory_ozone_wayland.h",
+          ]
+        }
       }
       if (is_linux) {
         sources += [ "style/platform_style_linux.cc" ]
@@ -1024,3 +1039,21 @@ if (is_mac) {
     ]
   }
 }
+
+if (use_ozone) {
+  action("generate_constructor_list") {
+    script = "../ozone/generate_constructor_list.py"
+    outputs = [
+      constructor_list_cc_file,
+    ]
+    platform_list_txt_file = "$target_gen_dir/../ozone/platform_list.txt"
+
+    args = [
+      "--platform_list=" + rebase_path(platform_list_txt_file, root_build_dir),
+      "--output_cc=" + rebase_path(constructor_list_cc_file, root_build_dir),
+      "--namespace=views",
+      "--typename=DesktopFactoryOzone",
+      "--include=\"ui/views/widget/desktop_aura/desktop_factory_ozone.h\"",
+    ]
+  }
+}
diff --git a/ui/views/DEPS b/ui/views/DEPS
index c26a0d2..da462d0 100644
--- a/ui/views/DEPS
+++ b/ui/views/DEPS
@@ -32,4 +32,7 @@ specific_include_rules = {
   "views_test_suite\.cc": [
     "+gpu/ipc/service",
   ]
+  "desktop_factory_ozone.\cc": [
+    "+ui/ozone"
+  ],
 }
diff --git a/ui/views/widget/desktop_aura/desktop_factory_ozone.cc b/ui/views/widget/desktop_aura/desktop_factory_ozone.cc
index e0a4489..1f7cc80 100644
--- a/ui/views/widget/desktop_aura/desktop_factory_ozone.cc
+++ b/ui/views/widget/desktop_aura/desktop_factory_ozone.cc
@@ -4,26 +4,36 @@
 
 #include "ui/views/widget/desktop_aura/desktop_factory_ozone.h"
 
+#include <memory>
+
 #include "base/logging.h"
+#include "ui/ozone/platform_object.h"
 
 namespace views {
 
 // static
-DesktopFactoryOzone* DesktopFactoryOzone::impl_ = NULL;
+DesktopFactoryOzone* DesktopFactoryOzone::impl_ = nullptr;
 
 DesktopFactoryOzone::DesktopFactoryOzone() {
+  DCHECK(!impl_) << "There should only be a single DesktopFactoryOzone.";
+  impl_ = this;
 }
 
 DesktopFactoryOzone::~DesktopFactoryOzone() {
+  DCHECK_EQ(impl_, this);
+  impl_ = nullptr;
 }
 
 DesktopFactoryOzone* DesktopFactoryOzone::GetInstance() {
-  CHECK(impl_) << "DesktopFactoryOzone accessed before constructed";
+  if (!impl_) {
+    std::unique_ptr<DesktopFactoryOzone> factory =
+        ui::PlatformObject<DesktopFactoryOzone>::Create();
+
+    // TODO(tonikitoo): Currently need to leak this object.
+    DesktopFactoryOzone* leaky = factory.release();
+    DCHECK_EQ(impl_, leaky);
+  }
   return impl_;
 }
 
-void DesktopFactoryOzone::SetInstance(DesktopFactoryOzone* impl) {
-  impl_ = impl;
-}
-
 } // namespace views
diff --git a/ui/views/widget/desktop_aura/desktop_factory_ozone.h b/ui/views/widget/desktop_aura/desktop_factory_ozone.h
index ea1c92e..fc795ee 100644
--- a/ui/views/widget/desktop_aura/desktop_factory_ozone.h
+++ b/ui/views/widget/desktop_aura/desktop_factory_ozone.h
@@ -27,9 +27,6 @@ class VIEWS_EXPORT DesktopFactoryOzone {
   // Returns the instance.
   static DesktopFactoryOzone* GetInstance();
 
-  // Sets the implementation delegate. Ownership is retained by the caller.
-  static void SetInstance(DesktopFactoryOzone* impl);
-
   // Delegates implementation of DesktopWindowTreeHost::Create externally to
   // Ozone implementation.
   virtual DesktopWindowTreeHost* CreateWindowTreeHost(
diff --git a/ui/views/widget/desktop_aura/desktop_factory_ozone_wayland.cc b/ui/views/widget/desktop_aura/desktop_factory_ozone_wayland.cc
new file mode 100644
index 0000000..b8f9386
--- /dev/null
+++ b/ui/views/widget/desktop_aura/desktop_factory_ozone_wayland.cc
@@ -0,0 +1,23 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ui/views/widget/desktop_aura/desktop_factory_ozone_wayland.h"
+
+namespace views {
+
+DesktopWindowTreeHost* DesktopFactoryOzoneWayland::CreateWindowTreeHost(
+    internal::NativeWidgetDelegate* native_widget_delegate,
+    DesktopNativeWidgetAura* desktop_native_widget_aura) {
+  return nullptr;
+}
+
+display::Screen* DesktopFactoryOzoneWayland::CreateDesktopScreen() {
+  return nullptr;
+}
+
+DesktopFactoryOzone* CreateDesktopFactoryOzoneWayland() {
+  return new DesktopFactoryOzoneWayland;
+}
+
+}  // namespace views
diff --git a/ui/views/widget/desktop_aura/desktop_factory_ozone_wayland.h b/ui/views/widget/desktop_aura/desktop_factory_ozone_wayland.h
new file mode 100644
index 0000000..844045d
--- /dev/null
+++ b/ui/views/widget/desktop_aura/desktop_factory_ozone_wayland.h
@@ -0,0 +1,41 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef UI_VIEWS_WIDGET_DESKTOP_AURA_DESKTOP_FACTORY_OZONE_WAYLAND_H_
+#define UI_VIEWS_WIDGET_DESKTOP_AURA_DESKTOP_FACTORY_OZONE_WAYLAND_H_
+
+#include "ui/views/views_export.h"
+#include "ui/views/widget/desktop_aura/desktop_factory_ozone.h"
+
+namespace display {
+class Screen;
+}
+
+namespace views {
+class DesktopNativeWidgetAura;
+class DesktopWindowTreeHost;
+
+namespace internal {
+class NativeWidgetDelegate;
+}
+
+class VIEWS_EXPORT DesktopFactoryOzoneWayland : public DesktopFactoryOzone {
+ public:
+
+  // Delegates implementation of DesktopWindowTreeHost::Create externally to
+  // Ozone implementation.
+  DesktopWindowTreeHost* CreateWindowTreeHost(
+      internal::NativeWidgetDelegate* native_widget_delegate,
+      DesktopNativeWidgetAura* desktop_native_widget_aura) override;
+
+  // Delegates implementation of DesktopScreen externally to
+  // Ozone implementation.
+  display::Screen* CreateDesktopScreen() override;
+};
+
+DesktopFactoryOzone* CreateDesktopFactoryOzoneWayland();
+
+}  // namespace views
+
+#endif // UI_VIEWS_WIDGET_DESKTOP_AURA_DESKTOP_FACTORY_OZONE_WAYLAND_H_
-- 
2.7.4

